Source/Core/Core/HW/GCMemcard/GCMemcard.cpp

```diff
@@ -27,196 +27,256 @@ static void ByteSwap(u8* valueA, u8* valueB)
+  error_code |= header_error_code;
+  if (header_error_code.HasCriticalErrors())
+    return std::make_pair(error_code, std::nullopt);
+
+  // The GC BIOS counts any card as corrupted as long as at least any two of [dir0, dir1, bat0,
+  // bat1] are corrupted. Yes, even if we have one valid dir and one valid bat, and even if those
+  // are both supposedly the newer ones.
+  //
+  // If both blocks of a single category are non-corrupted the used block depends on the update
+  // counter. If both blocks have the same update counter, it prefers block 0. Otherwise it prefers
+  // whichever block has the higher value. Essentially, if (0.update_ctr >= 1.update_ctr) { use 0 }
+  // else { use 1 }.
+  //
+  // If a single block of the four is corrupted, the non-corrupted one of the same category is
+  // immediately copied over the corrupted block with an incremented update counter. At this point
+  // both blocks contain the same data, so it's hard to tell which one is used, but presumably it
+  // uses the one with the now-higher update counter, same as it would have otherwise.
+  //
+  // This rule only applies for errors within a single block! That is, invalid checksums for both
+  // types, and free block mismatch for the BATs. Once two valid blocks have been selected but it
+  // later turns out they do not match eachother (eg. claimed block count of a file in the directory
+  // does not match the actual block count arrived at by following BAT), the card will be treated as
+  // corrupted, even if perhaps a different combination of the two blocks would result in a valid
+  // memory card.
+
+  // can return invalid checksum, data in unused area
+  GCMemcardErrorCode dir_block_0_error_code = card.m_directory_blocks[0].CheckForErrors();
+  GCMemcardErrorCode dir_block_1_error_code = card.m_directory_blocks[1].CheckForErrors();
+
+  // can return invalid card size, invalid checksum, data in unused area, free block mismatch
+  GCMemcardErrorCode bat_block_0_error_code = card.m_bat_blocks[0].CheckForErrors(card_size_mbits);
+  GCMemcardErrorCode bat_block_1_error_code = card.m_bat_blocks[1].CheckForErrors(card_size_mbits);
+
+  const bool dir_block_0_valid = !dir_block_0_error_code.HasCriticalErrors();
+  const bool dir_block_1_valid = !dir_block_1_error_code.HasCriticalErrors();
+  const bool bat_block_0_valid = !bat_block_0_error_code.HasCriticalErrors();
+  const bool bat_block_1_valid = !bat_block_1_error_code.HasCriticalErrors();
+
+  // if any two (at least) blocks are corrupted return failure
+  int number_of_corrupted_dir_bat_blocks = 0;
+  if (!dir_block_0_valid)
+    ++number_of_corrupted_dir_bat_blocks;
+  if (!dir_block_1_valid)
+    ++number_of_corrupted_dir_bat_blocks;
+  if (!bat_block_0_valid)
+    ++number_of_corrupted_dir_bat_blocks;
+  if (!bat_block_1_valid)
+    ++number_of_corrupted_dir_bat_blocks;
+
+  if (number_of_corrupted_dir_bat_blocks > 1)
```

Comment:
I wonder if we should allow this as potential data recovery option (when one dir/bat pair is still valid but the GC BIOS would reject it) - but this is probably out of scope here.

Reply:
I left a note in the code about potentially adding that later.